django-test-suite: startswith lookup + underscore values not working

```
======================================================================
FAIL: test_escaping (lookup.tests.LookupTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/tests/lookup/tests.py", line 453, in test_escaping
    ['<Article: Article_ with underscore>']
  File "/home/tim/code/django/django/test/testcases.py", line 1056, in assertQuerysetEqual
    return self.assertEqual(list(items), values, msg=msg)
AssertionError: Lists differ: [] != ['<Article: Article_ with underscore>']

Second list contains 1 additional elements.
First extra element 0:
'<Article: Article_ with underscore>'

- []
+ ['<Article: Article_ with underscore>']

======================================================================
FAIL: test_exclude (lookup.tests.LookupTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/tests/lookup/tests.py", line 509, in test_exclude
    '<Article: Article 1>',
  File "/home/tim/code/django/django/test/testcases.py", line 1056, in assertQuerysetEqual
    return self.assertEqual(list(items), values, msg=msg)
AssertionError: Lists differ: ['<Ar[90 chars]ticle_ with underscore>', '<Article: Article 5[142 chars] 1>'] != ['<Ar[90 chars]ticle 5>', '<Article: Article 6>', '<Article: [103 chars] 1>']

First differing element 2:
'<Article: Article_ with underscore>'
'<Article: Article 5>'

First list contains 1 additional elements.
First extra element 9:
'<Article: Article 1>'

  ['<Article: Article with \\ backslash>',
   '<Article: Article% with percent sign>',
-  '<Article: Article_ with underscore>',
   '<Article: Article 5>',
   '<Article: Article 6>',
   '<Article: Article 4>',
   '<Article: Article 2>',
   '<Article: Article 3>',
   '<Article: Article 7>',
   '<Article: Article 1>']
```

--- a/spanner/django/base.py
+++ b/spanner/django/base.py
@@ -69,14 +69,21 @@ class DatabaseWrapper(BaseDatabaseWrapper):
     operators = {
         'exact': '= %s',
         'iexact': 'REGEXP_CONTAINS(%s, %%%%s)',
-        'contains':  'LIKE %s',
+        # contains uses REGEXP_CONTAINS instead of LIKE to allow
+        # DatabaseOperations.prep_for_like_query() to do regular expression
+        # escaping. prep_for_like_query() is called for all the lookups that
+        # use REGEXP_CONTAINS except regex/iregex (see
+        # django.db.models.lookups.PatternLookup).
+        'contains':  'REGEXP_CONTAINS(%s, %%%%s)',
         'icontains': 'REGEXP_CONTAINS(%s, %%%%s)',
         'gt': '> %s',
         'gte': '>= %s',
         'lt': '< %s',
         'lte': '<= %s',
-        'startswith': 'STARTS_WITH(%s, %%%%s)',
-        'endswith': 'ENDS_WITH(%s, %%%%s)',
+        # Using REGEXP_CONTAINS instead of STARTS_WITH and ENDS_WITH for the
+        # same reasoning as described above for 'contains'.
+        'startswith': 'REGEXP_CONTAINS(%s, %%%%s)',
+        'endswith': 'REGEXP_CONTAINS(%s, %%%%s)',
         'istartswith': 'REGEXP_CONTAINS(%s, %%%%s)',
         'iendswith': 'REGEXP_CONTAINS(%s, %%%%s)',
         'regex': 'REGEXP_CONTAINS(%s, %%%%s)',

--- a/spanner/django/features.py
+++ b/spanner/django/features.py
@@ -81,8 +81,4 @@ class DatabaseFeatures(BaseDatabaseFeatures):
         # 'DatabaseWrapper' object has no attribute 'pattern_ops'
         # https://github.com/orijtech/spanner-orm/issues/178
         'lookup.tests.LookupTests.test_pattern_lookups_with_substr',
-        # startswith lookup + underscore values not working
-        # https://github.com/orijtech/spanner-orm/issues/179
-        'lookup.tests.LookupTests.test_escaping',
-        'lookup.tests.LookupTests.test_exclude',
     )

--- a/spanner/django/lookups.py
+++ b/spanner/django/lookups.py
@@ -12,22 +12,24 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import re
-
 from django.db.models.lookups import (
-    EndsWith, IContains, IEndsWith, IExact, IRegex, IStartsWith, Regex,
-    StartsWith,
+    Contains, EndsWith, IContains, IEndsWith, IExact, IRegex, IStartsWith,
+    Regex, StartsWith,
 )
 
 
-def icontains(self, compiler, connection):
+def contains(self, compiler, connection):
+    """contains and icontains"""
     lhs_sql, params = self.process_lhs(compiler, connection)
     rhs_sql, rhs_params = self.process_rhs(compiler, connection)
     params.extend(rhs_params)
     rhs_sql = self.get_rhs_op(connection, rhs_sql)
     # Chop the leading and trailing percent signs that Django adds to the
     # param since this isn't a LIKE query as Django expects.
-    params[0] = '(?i)%s' % re.escape(params[0][1:-1])
+    params[0] = params[0][1:-1]
+    # Add the case insensitive flag for icontains.
+    if self.lookup_name.startswith('i'):
+        params[0] = '(?i)' + params[0]
     # rhs_sql is REGEXP_CONTAINS(%s, %%s), and lhs_sql is the column name.
     return rhs_sql % lhs_sql, params
 
@@ -38,7 +40,7 @@ def iexact(self, compiler, connection):
     params.extend(rhs_params)
     rhs_sql = self.get_rhs_op(connection, rhs_sql)
     # Wrap the parameter in ^ and $ to restrict the regex to an exact match.
-    params[0] = '^(?i)%s$' % re.escape(params[0])
+    params[0] = '^(?i)%s$' % params[0]
     # rhs_sql is REGEXP_CONTAINS(%s, %%s), and lhs_sql is the column name.
     return rhs_sql % lhs_sql, params
 
@@ -66,20 +68,19 @@ def startswith_endswith(self, compiler, connection):
     # Chop the leading (endswith) or trailing (startswith) percent sign that
     # Django adds to the param since this isn't a LIKE query as Django expects.
     if 'endswith' in self.lookup_name:
-        params[0] = params[0][1:]
+        params[0] = str(params[0][1:]) + '$'
     else:
-        params[0] = params[0][:-1]
-    # Construct the regex for istartswith or iendswith.
+        params[0] = '^' + str(params[0][:-1])
+    # Add the case insentiive flag for istartswith or iendswith.
     if self.lookup_name.startswith('i'):
-        regex = '^(?i)%s' if self.lookup_name == 'istartswith' else '(?i)%s$'
-        params[0] = regex % re.escape(params[0])
-    # rhs_sql is e.g. 'STARTS_WITH(%s, %%s)' or REGEXP_CONTAINS(%s, %%s), and
-    # lhs_sql is the column name.
+        params[0] = '(?i)' + params[0]
+    # rhs_sql is REGEXP_CONTAINS(%s, %%s), and lhs_sql is the column name.
     return rhs_sql % lhs_sql, params
 
 
 def register_lookups():
-    IContains.as_spanner = icontains
+    Contains.as_spanner = contains
+    IContains.as_spanner = contains
     IExact.as_spanner = iexact
     Regex.as_spanner = regex
     IRegex.as_spanner = regex

--- a/spanner/django/operations.py
+++ b/spanner/django/operations.py
@@ -1,3 +1,4 @@
+import re
 from base64 import b64decode
 from datetime import datetime
 from decimal import Decimal
@@ -141,3 +142,9 @@ class DatabaseOperations(BaseDatabaseOperations):
                            'endswith', 'iendswith', 'regex', 'iregex'):
             return 'CAST(%s AS STRING)'
         return '%s'
+
+    def prep_for_like_query(self, x):
+        """Lookups that use this method use REGEX_CONTAINS instead of LIKE."""
+        return re.escape(str(x))
+
+    prep_for_iexact_query = prep_for_like_query

